# Cleaning the Orderlines dataframe

Read in data

In [22]:
import pandas as pd

In [ ]:
path = '../data/data_raw/'
orderlines_orig = pd.read_csv(path + 'orderlines.csv')

In [24]:
ol = orderlines_orig.copy()

### Check for duplicates

In [25]:
ol.duplicated().sum()

0

In [36]:
ol['id'].duplicated().sum()

0

### Eyeball the data

In [26]:
ol.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


In [27]:
ol.isna().sum()

id                  0
id_order            0
product_id          0
product_quantity    0
sku                 0
unit_price          0
date                0
dtype: int64

In [28]:
ol.describe()

,id,id_order,product_id,product_quantity
count,2.939830e+05,293983.000000,293983.0,293983.000000
mean,1.397918e+06,419999.116544,0.0,1.121126
std,1.530096e+05,66344.486479,0.0,3.396569
min,1.119109e+06,241319.000000,0.0,1.000000
25%,1.262542e+06,362258.500000,0.0,1.000000
50%,1.406940e+06,425956.000000,0.0,1.000000
75%,1.531322e+06,478657.000000,0.0,1.000000
max,1.650203e+06,527401.000000,0.0,999.000000


- unit_price and date need to be reformatted to numeric and datetime, respectively
- max product_quantity is 999 - seems strange (but this is maybe from an order that is cancelled or still in shopping basket - investigate in Data Quality step)

### Reformat unit_price to numeric

In [29]:
#remove first '.' in unit_price
dot_mask = ol['unit_price'].str.count('\.')==2
ol.loc[dot_mask, 'unit_price'] = ol.loc[dot_mask, 'unit_price'].str.replace('.', '', 1)
ol['unit_price_num'] = pd.to_numeric(ol['unit_price'])


In [30]:
ol.sample(10)

,id,id_order,product_id,product_quantity,sku,unit_price,date,unit_price_num
94423,1299779,380815,0,1,MOS0207,36.99,2017-07-22 19:49:03,36.99
293950,1650157,527372,0,1,APP0698,9.99,2018-03-14 13:41:18,9.99
4283,1132026,303500,0,1,JMO0091,32.90,2017-01-06 09:19:03,32.90
80945,1275409,369233,0,1,MOS0138,36.99,2017-06-26 17:05:47,36.99
13330,1151077,311516,0,1,SAT0044,19.99,2017-01-18 11:45:17,19.99
188522,1477678,455630,0,1,APP0665,82.99,2017-12-05 12:38:58,82.99
172484,1454898,446758,0,1,APP1089,213.99,2017-11-27 00:55:29,213.99
161437,1436421,438772,0,1,OSM0007,98.99,2017-11-24 12:26:18,98.99
206606,1504550,466596,0,1,APP2479,793.00,2017-12-19 11:08:58,793.00
4182,1131828,303399,0,1,JAW0048,29.99,2017-01-05 22:58:36,29.99


In [31]:
ol.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 8 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   id                293983 non-null  int64  
 1   id_order          293983 non-null  int64  
 2   product_id        293983 non-null  int64  
 3   product_quantity  293983 non-null  int64  
 4   sku               293983 non-null  object 
 5   unit_price        293983 non-null  object 
 6   date              293983 non-null  object 
 7   unit_price_num    293983 non-null  float64
dtypes: float64(1), int64(4), object(3)
memory usage: 17.9+ MB


In [32]:
ol.isna().sum()

id                  0
id_order            0
product_id          0
product_quantity    0
sku                 0
unit_price          0
date                0
unit_price_num      0
dtype: int64

In [33]:
ol['unit_price_num'].describe()

count    293983.000000
mean        409.855192
std         837.065461
min        -119.000000
25%          35.990000
50%          92.990000
75%         339.580000
max      159989.830000
Name: unit_price_num, dtype: float64

Min value (-119) and max value (159989) look strange
(possibly these orders are from cancelled/shopping cart orders - deal with them in Data Quality step)

### Reformat date as datetime

In [34]:
ol['date_dt'] = pd.to_datetime(ol['date'])
ol.sample(10)

,id,id_order,product_id,product_quantity,sku,unit_price,date,unit_price_num,date_dt
89988,1291547,376813,0,1,PAC1214,652.99,2017-07-13 11:46:27,652.99,2017-07-13 11:46:27
205477,1502309,465570,0,1,LAC0190,261.79,2017-12-18 12:07:21,261.79,2017-12-18 12:07:21
113720,1346726,397277,0,1,APP1719,469.00,2017-09-07 02:21:00,469.00,2017-09-07 02:21:00
205749,1502842,465816,0,1,MOS0239,19.99,2017-12-18 15:45:19,19.99,2017-12-18 15:45:19
92832,1296788,379401,0,2,SEA0110,134.99,2017-07-19 09:43:04,134.99,2017-07-19 09:43:04
35811,1195313,330818,0,1,LAC0224,381.65,2017-03-01 08:10:32,381.65,2017-03-01 08:10:32
246008,1574197,496147,0,1,APP2140,1419.00,2018-01-22 15:36:33,1419.00,2018-01-22 15:36:33
15169,1154736,313023,0,1,APP0666,264.99,2017-01-21 21:37:46,264.99,2017-01-21 21:37:46
284361,1635921,521379,0,2,APP2288,2985.00,2018-03-05 10:36:05,2985.00,2018-03-05 10:36:05
89004,1289760,375967,0,1,SAT0029,42.99,2017-07-11 16:54:57,42.99,2017-07-11 16:54:57


In [35]:
ol.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 9 columns):
 #   Column            Non-Null Count   Dtype         
---  ------            --------------   -----         
 0   id                293983 non-null  int64         
 1   id_order          293983 non-null  int64         
 2   product_id        293983 non-null  int64         
 3   product_quantity  293983 non-null  int64         
 4   sku               293983 non-null  object        
 5   unit_price        293983 non-null  object        
 6   date              293983 non-null  object        
 7   unit_price_num    293983 non-null  float64       
 8   date_dt           293983 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(1), int64(4), object(3)
memory usage: 20.2+ MB


looks good!

### Save to csv

In [37]:
ol.to_csv('data_cleaned/orderlines_cleaned.csv')